### PR TITLE
Better extensions for sum/average, deprecate others

### LINKED ISSUE #28: MathFluxExtensions Missing Methods, and Inconsistent
## Motivation

The use of Kotlin Math Extensions is limited as it is missing several methods specific to sum() and average() of typed fluxes such as BigDecimal.  I write financial application(s) which uses BigDecimal as a preference, using floats and doubles is discouraged, hence the need for Big Decimal specific implementations. 

As such I have to implement these missing functions in my own Kotlin code, which defeats the purpose of this library. Traversing via Double is not an option.

### Implement missing methods

The following is an example of a missing method, of which there are many examples. 

```
fun <T> Flux<T>.sumBigDecimal(mapper: (T) -> Number): Mono<BigDecimal>
        = MathFlux.sumBigDecimal(this, Function(mapper)) 
```

Ideally there would be a one - to - one relationship between the Kotlin functions provided by this library and the functions they wrap in the core MathFlux  implementation.

### Untyped methods

In the library there are a few examples of untyped methods() e.g.

```
fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
        = MathFlux.sumLong(this, Function(mapper)) 
```

and

```
fun <T> Flux<T>.average(mapper: (T) -> Number): Mono<Double>
        = MathFlux.averageDouble(this, Function(mapper)) 
```

These untyped method make (opinions) on the underlying type that should be used. These opinions are also not consistent, since one method uses Long the other Double, unsure why this is. 

Noting also that the underlying library MathFlux provides no such opinions, this is specific to the Kotlin Extension.

Could do one of the following.

(a) Deprecate the use of these untyped methods
(b) Make the untyped methods consistent e.g. Double (an opinion)
(c) Do Nothing. Leave the untyped methods as they are.

## Considered alternatives
N/A

## Additional context

PS. I am happy to submit a Pull Request if you like.  Would need to understand the Scope of the change as discussed above, happy to take any direction on this.

PPS. This is a continuation of an issue I am trying to get resolved Part 1 was fixed in this issue.
https://github.com/reactor/reactor-addons/issues/260

### PR DESCRIPTION
**Pull request by kiwisincebirth**
fix #28 

Added missing Kotlin wrappers for all sum() and average() methods provided in reactor.math.MathFlux to the library

 - Added unit test to cover all of these methods using the existing unit tests as a guide
 - Include Java Doc for all new methods with @ since based on the current snapshot version. 
 - Note Inconsistency, since the existing methods have @ since based on reactor-extra version numbering.

### CODE REVIEW

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 1]
```diff
@@ -33,6 +34,45 @@ import java.util.function.Function
...
+ * and return it as a [Mono] of [Int].
+ *
+ * Note that numbers will be mapped to [Int] using Java standard conversions
+ * This may lead to arbitrary precision gain or loss if the source flux contains different types
+ * Please choose an appropriate method based on the expected types in the source flux.
+ *
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T: Number> Flux<T>.sumInt(): Mono<Int> = MathFlux.sumInt(this)
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: would there be compiler ambiguity if we kept `<T:Number> sum` above AND turn these additional methods into `<T:Int> sum`, `<T:Long> sum`, etc ?
my goal was to use Kotlin type system as much as possible to avoid the "dumb" style of the Java `MathFlux` (which repeats the type in the method name).

Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: Totally get where you are coming from. Your suggestion Does create a compiler Conflict with the existing `sum(): Mono<Long>` method

However there is potentially a workable solution which achieves the outcome. But it requires replacing the existing `sum(): Mono<Long>` method with the below method.

```
inline fun <reified T: Number> Flux<T>.sum(): Mono<T> =
    when (T::class) {
        BigDecimal::class -> MathFlux.sumBigDecimal(this) as Mono<T>
        Double::class -> MathFlux.sumDouble(this) as Mono<T>
        Long::class -> MathFlux.sumLong(this) as Mono<T>
        // todo above cases need to be exhaustive
        else -> Mono.empty() // or error()
    }
```

Disadvantages 
1) Existing users of sum() may be broken since new function returns Mono T (not Mono Long ) - so existing code would would need to explicitly migrate to sumLong() - OR NOT - depending on use of resultant Mono<T>

Advantages
1) The sum() method would work for all cases (probably majority) where the input types holds enough precision, and there is no expectation of type conversion.
2) The new sum() would always return the same type that was passed in so there is no ambiguity. eg. `fluxOfInts.sum()` - makes no statement about output type so user can assume same type.
3) If the user needs to increase precision or change output type then the full method e.g. e.g. `fluxOfInts.sumLong()` - makes an Explicit statement about the (output) type of result that is needed.

Comment (simonbasle) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: I very much like that 🤩

Perhaps the same style could be applied to the map-and-sum/average case ?
And there, since the mapping is explicit, no need for individual per-type methods at all!

Regarding the disadvantage, even if I'd argue that it is more desirable for the `sum` method to maintain the type, it is indeed a breaking compatibility change 😞

can we get creative with the name and deprecate the old `sum()` perhaps?

in Reactor we have the `collect()` and `reduce()` operators.
So perhaps something like `collectSum()`/`collectAverage()`?

For the map case, `mapAndCollectSum()` and `mapAndCollectAverage()`?

Any good candidates for names from the Kotlin world?

Comment (simonbasle) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]:
```java
Flux<Int> fluxOfInts;
Mono<Int> sumInt = fluxOfInts.collectSum();
Mono<Long> sumLong = fluxOfInts.sumLong();
Mono<Int> avgInt = fluxOfInts.collectAverage();
Mono<Double> avgInt = fluxOfInts.averageDouble();

Flux<String> fluxOfSentences;
Mono<Double> numberOfCharsDouble = fluxOfSentences.mapAndCollectSum(s -> (double) s.length());
Mono<Int> numberOfChars = fluxOfSentences.mapAndCollectSum(String::length)
Mono<Int> averageWordsPerSentence = fluxOfSentences.mapAndCollectAverage(s -> s.split(" ").length);
```

Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: > Perhaps the same style could be applied to the map-and-sum/average case ?

Absolutely that was the intent.

> can we get creative with the name and deprecate the old sum() perhaps?

or make the breaking change in a major version release.

> can we get creative with the name and deprecate the old sum() perhaps?

Deprecation would be a good idea in any case.

> So perhaps something like collectSum()/collectAverage()?

Fo these methods I dont have any better idea's than what you suggest

> Any good candidates for names from the Kotlin world?

In kotlin world found the sumOf() method in sequence's where a mapping function exists, suggest this would be reasonable alternative, and is shorter.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/sum-of.html

thus
```
Flux<String> fluxOfSentences;
Mono<Double> numberOfCharsDouble = fluxOfSentences.sumOf(s -> (double) s.length());
Mono<Int> numberOfChars = fluxOfSentences.sumOf(String::length)
Mono<Int> averageWordsPerSentence = fluxOfSentences.averageOf(s -> s.split(" ").length);
```

As for the non-mapping functions, couple of other idea's 

* sumAny() - Any being an abstract type, but it adds unnecessary type to method name 
* sumOf() - overloaded - same as mapping function - not particularly good.
* sumAll() 
* sumJust() - `Just` implies just the unmapped object.
* sumFlux() 
* sumItems() 
* sumTerms()
* sumToMono()

I could come up with many more suggestions none particularly compelling  - when really just want to use `sum()` as is the case with JDK and Kotlin collections frameworks - An advantage with `sum` appearing first, may be more discoverable (alpha sorting), more easily found in any autocomplete, `collect` is harder to find (IMHO)

`sumToMono()` is slightly more interesting as an untyped method if you pair it with typed version(s) such as `sumToDouble()` - then you have naming consistency. Just another random idea.

Happy to go with your suggestion `collectSum()` until future major release where the deprecated sum() average() functions potentially could be replaced.

**Comment by simonbasle** [comment]
Sounds like a plan! collectX, xOf and deprecation

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 1 above)
Comment (simonbasle) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: Ah, sumAll might be a good one too

**Comment by kiwisincebirth** [comment]
> Sounds like a plan! collectX, xOf and deprecation

> Ah, sumAll might be a good one too

Can you clarify your preference, based on comments from above ? 

(a) collectX() and xOf()
(b) xAll() and xOf()
(c) xAll() - Noting there is no method signature conflict between mapped and unmapped functions, so names can be identical. we can just have sumAll() and averageAll()
(d) something else

Option (b or c) would be my preference.

PS. Have committed the new methods with deprecations, except naming is not correct at this time. I was more exhaustive in the switch case statements, to account for the most common data types to avoid runtime errors.

**Comment by simonbasle** [comment]
> (a) collectX() and xOf()
> (b) xAll() and xOf()
> (c) xAll() - Noting there is no method signature conflict between mapped and unmapped functions, so names can be identical. we can just have sumAll() and averageAll()
> (d) something else

hehe, so much choice... actually, you have a good point with (c). `sumAll()` is merely a simplified case of `sumAll(Function<T, R:Number>)` where `T` is already a `Number` sub-class.

to summarize:

 - `sumAll()` and `averageAll()` would maintain the original `T:Number` type
 - `sumAll(T -> N:Number)` and `averageAll(T -> N:Number)` would explicitly convert from any type `T` to `N` Number sub-type (so no ambiguity in the output type here)
 - `sum()`/`average()`/`sum(Function)`/`average(Function)` get deprecated

Do we still need `sumLong...sumBigDecimal` and all ?

**Comment by kiwisincebirth** [comment]
Have made the changes to the code, specifically to name methods as sumAll() and averageAll()

> Do we still need sumLong...sumBigDecimal and all ?

I believe so - consider the following code

```
    val fluxOfInts: Flux<Int> = Flux.just(1,2,3,4,5)
    val sumLong = fluxOfInts.sumLong()
    val avgDouble = fluxOfInts.averageDouble()
```

if you remove the typed functions the above code would have to be written as

```
    val fluxOfInts: Flux<Int> = Flux.just(1,2,3,4,5)
    val sumLong = fluxOfInts.sumAll(kotlin.Int::toLong)
    val avgDouble = fluxOfInts.averageAll(kotlin.Int::toDouble)
```

Which I think is good reason to keep the typed functions.

**Comment by simonbasle** [comment]
> Which I think is good reason to keep the typed functions.

One last naming nitpick/proposal then: what do you think about `sumAsXxx` to better convey that a conversion is potentially in play? (`sumAsLong`, `sumAsDouble`, etc...)

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 2]
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
  * unexpected averages.
  *
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]:
```suggestion
 * @deprecated Please use averageDouble() as a direct replacement or consider more general purpose averageAll().
 * To be removed at the earliest in 1.3.0.
```

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 3]
```diff
@@ -86,10 +286,57 @@ fun <T: Comparable<T>> Flux<T>.max(): Mono<T> = MathFlux.max(this)
  *
  * @param mapper a lambda converting values to [Number]
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]:
```suggestion
 * @deprecated Please use sumAll(mapper) instead. To be removed in 1.3.0 at the earliest.
```

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 4]
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
  * unexpected averages.
  *
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
  */
 fun <T: Number> Flux<T>.average(): Mono<Double> = MathFlux.averageDouble(this)
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:207]:
```suggestion
@Deprecated(message = "Please use averageDouble() as a direct replacement or consider more general purpose averageAll()",
replaceWith = ReplaceWith("averageDouble()"))
fun <T: Number> Flux<T>.average(): Mono<Double> = MathFlux.averageDouble(this)
```

Comment (simonbasle) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:207]: missing `@Deprecated` annotation. I'm not entirely sure of the syntax and the annotation might need to be imported, as I'm suggesting this from Github's UI.

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 5]
```diff
@@ -86,10 +286,57 @@ fun <T: Comparable<T>> Flux<T>.max(): Mono<T> = MathFlux.max(this)
  *
  * @param mapper a lambda converting values to [Number]
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
  */
 fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:319]:
```suggestion
@Deprecated(message = "Please use sumAll(mapper) instead",
replaceWith = ReplaceWith("sumAll(mapper)"))
fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
```

Comment (simonbasle) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:319]: same as above

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 6]
```diff
@@ -29,10 +56,50 @@ import java.util.function.Function
  * Note that summing decimal numbers with this method loses precision, see [sumDouble].
  *
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]:
```suggestion
 * @deprecated Please use sumLong() as a direct replacement or consider more general purpose sumAll().
To be removed at the earliest in 1.3.0.
```

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 7]
```diff
@@ -29,10 +56,50 @@ import java.util.function.Function
...
  */
 fun <T: Number> Flux<T>.sum(): Mono<Long> = MathFlux.sumLong(this)
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:66]:
```suggestion
@Deprecated(message = "Please use sumLong() as a direct replacement or consider more general purpose sumAll()",
replaceWith = ReplaceWith("sumLong()"))
fun <T: Number> Flux<T>.sum(): Mono<Long> = MathFlux.sumLong(this)
```

**Comment by simonbasle** [comment]
oh and while I agree the `sumType()` methods have value, I really believe the `sumType(Function)` and `averageType(Function)` ones should be removed (since `sumAll(Function)`/`averageAll(Function)` would be expressive and concise enough to cover these cases).

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 8]
```diff
@@ -86,10 +286,57 @@ fun <T: Comparable<T>> Flux<T>.max(): Mono<T> = MathFlux.max(this)
...
+ *
+ * Note that numbers will be mapped to [Int] using Java standard conversions
+ * This may lead to arbitrary precision gain or loss if the source flux contains different types
+ * Please choose an appropriate method based on the expected types in the source flux.
+ *
+ * @param mapper a lambda converting values to [Number]
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T> Flux<T>.sumInt(mapper: (T) -> Number): Mono<Int>
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: these methods with number suffix and Function param should go

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 9]
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
  *
  * @param mapper a lambda converting values to [Number]
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]:
```suggestion
 * @deprecated Please use averageAll(mapper) instead
```

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 10]
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
...
  */
 fun <T> Flux<T>.average(mapper: (T) -> Number): Mono<Double>
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:389]:
```suggestion
@Deprecated(message = "Please use averageAll(mapper) instead",
replaceWith = ReplaceWith("averageAll(mapper)"))
fun <T> Flux<T>.average(mapper: (T) -> Number): Mono<Double>
```

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 11]
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
...
 
 //sum/sumDouble/average lambda versions where a converter is provided
+
+/**
+ * General purpose extension function to map arbitrary values in a [Flux] to [Number]s and
+ * return the sum of these Numbers as a [Mono] of [Number].
+ * The resultant [Mono] will have the same [Number] type as the output of the mapping function
+ *
+ * If the result type (or precision) needs to be changed then a specific sumType() method
+ * should be used in preference to this method. e.g. sumDouble(mapper) can be used to compute
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: as commented on the PR level, I think `sumDouble(mapper)` shouldn't be included.
either the `T` is not a number, and the mapper should directly extract to `Double`, or it is a number and thus using `sumAll(mapper)` doesnt't make sense, one can directly use `sumDouble()` (without a mapper).

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 12]
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
  *
  * @param mapper a lambda converting values to [Number]
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:333]: this one should actually be deprecated like the others (since it pre-dates the PR)

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 13]
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
...
+fun <T> Flux<T>.sumBigDecimal(mapper: (T) -> Number): Mono<BigDecimal>
+        = MathFlux.sumBigDecimal(this, Function(mapper))
+
+/**
+ * General purpose extension function to map arbitrary values in a [Flux] to [Number]s and
+ * return the average of these Numbers as a [Mono] of [Number].
+ * The resultant [Mono] will have the same [Number] type as the output of the mapping function
+ *
+ * If the result type (or precision) needs to be changed then a specific averageType() method
+ * should be used in preference to this method. e.g. averageDouble(mapper) can be used to compute
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: as in `sumAll(mapper)`, this shouldn't reference `averageDouble(mapper)` but simply `averageDouble()`

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 14]
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
...
+ * @since 1.1.5
+ */
+@Suppress("UNCHECKED_CAST")
+inline fun <T: Any, reified R: Number> Flux<T>.averageAll(noinline mapper: (T) -> R): Mono<R> =
+    when (R::class) {
+        BigDecimal::class -> MathFlux.averageBigDecimal(this, Function(mapper)) as Mono<R>
+        BigInteger::class -> MathFlux.averageBigInteger(this, Function(mapper)) as Mono<R>
+        Double::class -> MathFlux.averageDouble(this, Function(mapper)) as Mono<R>
+        Float::class -> MathFlux.averageFloat(this, Function(mapper)) as Mono<R>
+        Long::class -> MathFlux.averageBigInteger(this, Function(mapper))
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:362]: I assume this is for better precision of the _average_, an issue that doesn't occur in sums?

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 15]
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
...
+ *
+ * Note that numbers will be mapped to [Float] using Java standard conversions
+ * This may lead to arbitrary precision gain or loss if the source flux contains different types
+ * Please choose an appropriate method based on the expected types in the source flux.
+ *
+ * @param mapper a lambda converting values to [Number]
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T> Flux<T>.averageFloat(mapper: (T) -> Number): Mono<Float>
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: as for sumTypes(mapper), averageTypes(mapper) should not be included in this PR

**Review by simonbasle — CHANGES_REQUESTED** [review]
refining the javadoc, missing deprecated annotations (with replaceWith) and commented on the first of each typed mapper-based variant that these shouldn't be included. we're almost there @kiwisincebirth this is looking better and better 👍

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 15 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: Have removed these methods.

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 2 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: Resolved in commit

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 3 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: addressed in commit.

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 4 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:207]: Added missing annotations

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 6 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: resolved in commit

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 8 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: resolved

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 10 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:389]: resolved

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 11 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: addressed

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 12 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:333]: all original methods are now deprecated

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 14 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:362]: Good pickup.

It is actually because the underlying Library does not support the averaging of integer type numbers, However for some reason BigInteger is supported for average.

Is this an omission - sort of IMHO - While most people think of averages and decimal numbers (obvious reasons), an average expressed as an integer is still an average (just with less precision) 

For completeness am using Big integer to do the averages then map the result back to the requested type.

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 13 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: resolved

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 9 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: I have improved the documentation but kept averageAsDouble() as the replacement

**Comment by kiwisincebirth** [comment]
First sorry for not resolving the tests before prior commit. 

> One last naming nitpick/proposal then: what do you think about sumAsXxx to better convey that a conversion is potentially in play? (sumAsLong, sumAsDouble, etc...)

Yes I like that have changed to include 'As' in all typed methods i.e. sumAsType(), averageAsType()

> oh and while I agree the sumType() methods have value, I really believe the sumType(Function) and averageType(Function) ones should be removed (since sumAll(Function)/averageAll(Function) would be expressive and concise enough to cover these cases).

Yep you are right on that, have removed the typed methods with function parameters, that I introduced, cannot remove sumDouble(Function) as it is part of the library

> missing @Deprecated annotation. I'm not entirely sure of the syntax and the annotation might need to be imported, as I'm suggesting this from Github's UI.

Added these annotations to the deprecated methods.

Have been through all the comments from above and marked resolved everything where I fixed. See response to your assumption above. 

Will commit shortly.

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 16]
```diff
@@ -57,8 +58,11 @@ inline fun <reified T:Number> Flux<T>.sumAll(): Mono<T> =
  *
  * @author Simon Baslé
  * @since 1.0.0
- * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
+ * @deprecated Please use sumLong() as a direct replacement, or consider more general purpose sumAll().
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: dang. it's now `sumAsLong` isn't it? deprecated messages need to be updated (doc, annotation message and annotation replaceWith) here and in the other similar deprecations

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt [hunk 17]
```diff
@@ -178,8 +199,11 @@ inline fun <reified T:Number> Flux<T>.averageAll(): Mono<T> =
  *
  * @author Simon Baslé
  * @since 1.0.0
- * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
+ * @deprecated Please use averageDouble() as a direct replacement,
```
Comment (simonbasle) [inline src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: update method name here too

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 17 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: resolved

File: src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt (hunk 16 above)
Comment (kiwisincebirth) [reply src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt]: resolved

**Comment by kiwisincebirth** [comment]
resolved the 2 issues 🤞

**Review by simonbasle — APPROVED** [review]
thanks for the good work @kiwisincebirth 👍

**Comment by kiwisincebirth** [comment]
pleasure.
